tools/deploy_firmware.py

```diff
@@ -0,0 +1,287 @@
+            print("Connecting to serial")
+            shell = pexpect.fdpexpect.fdspawn(ser, logfile=sys.stdout.buffer)
+            if not shell.isalive():
+                raise ValueError("Unable to connect to the serial device!")
+            print("Connected")
+            # The console might not be active yet:
+            shell.sendline("")
+            # make the shell prompt appear:
+            self.set_prompt(shell)
+            shell.expect(self.serial_prompt)
+            # Do the upgrade
+            shell.sendline("sysupgrade -v /tmp/{}".format(self.image))
+            # first give it 30 seconds, and fail early if the upgrade didn't start:
+            shell.expect(r"Performing system upgrade\.\.\.", timeout=30)
+            # now give it more time to apply the upgrade and reboot:
+            shell.expect(r"Rebooting system\.\.\.", timeout=300)
+            shell.expect("Please press Enter to activate this console", timeout=180)
+            # activate the console:
+            shell.sendline("")
+            self.set_prompt(shell)
+            shell.expect(self.serial_prompt)
+            shell.sendline("exit")
+        if not self.reach(attempts=10):
+            raise ValueError("The device was not reachable after the upgrade!")
+
+
+def main():
+    parser = argparse.ArgumentParser(prog=sys.argv[0],
+                                     description="""Update a prplWrt device, either through u-boot
+                                     or using sysupgrade, depending on the target device.""")
+    parser.add_argument('-d', '--device',
+                        help="""Device to upgrade. Currently supported targets are: netgear-rax40
+                        glinet-b1300 turris-omnia""", required=True)
+    parser.add_argument(
+        '-t',
+        '--target-name',
+        help="Name of the target to upgrade (make sure it's reachable through ssh).", required=True)
+
+    parser.add_argument(
+        '-i',
+        '--image',
+        help="Name of the image to use for the upgrade (should exist in the artifacts folder).",
+        required=True)
+
+    args = parser.parse_args()
+
+    if args.device == "netgear-rax40":
+        dev = NetgearRax40(args.device, args.target_name, args.image)
+    else:
+        dev = Generic(args.device, args.target_name, args.image)
+
+    print("Checking if the device is reachable over ssh")
+    if not dev.reach():
+        raise ValueError("The device {} is not reachable over ssh! check your ssh configuration."
+                         .format(dev.name))
+    print("Checking if the device needs to be upgraded")
+    if dev.needs_upgrade():
+        print("The device {} will be upgraded".format(dev.name))
+        try:
+            dev.sysupgrade()
+        except NotImplementedError:
+            dev.upgrade_uboot()
+        print("Done")
+    else:
+        print("The device is already using the same version, no upgrade will be done.")
```

Comment:
I wonder if `force` option should be added, for cases that we might want to flash regardless of the current version.
What do you think?

Reply:
It could be useful if you want to try experimental builds I guess, but I'll keep that for later also.